### PR TITLE
[utils] Add support for different database types

### PR DESCRIPTION
**Pull request by s10**
[utils] Add support for different database types

Add an option to choose database type and generate different configurations
for ProxySQL and oslo.db connection

The following options are available:
- mariadb
- pxc-db

To choose the different database type
- Variable dbType should be set
- Chart (pxc-db or mariadb) should be present in the dependencies

To make this selection possible following things changed:
- Added helper function utils.db_host
- Added helper function utils.db_url
- Extended helper function db_credentials
- Added helpers `utils._db_url_*` and `utils._db_host_*` for mariadb and pxc-db

One specific case is supported which doesn't require a chart dependency:
```
proxysql:
  force_enable:
    - mariadb
```
This is currently being used in the Placement chart, because
Placement reuses nova_api database instance

Other changes in this commit:
- Improve readability of the db_host_pxc and db_url_pxc functions
- Update proxysql sidecar docs

### CODE REVIEW

File: openstack/utils/templates/_hosts.tpl [hunk 1]
```diff
@@ -102,12 +153,70 @@ mysql+pymysql://{{ include "db_credentials" . }}@
...
+    {{- if kindIs "map" . }}
+        {{- if and .Values.pxc_db.users .Values.pxc_db.databases }}
+            {{- $db := first .Values.pxc_db.databases }}
+            {{- $user := get .Values.pxc_db.users $db | required (printf ".Values.pxc_db.%v.name & .password are required (key comes from first database in .Values.pxc_db.databases)" $db) }}
+            {{- tuple . $db $user.name (required (printf "User with key %v requires password" $db) $user.password) | include "utils._db_url_pxc_db_v2" }}
+        {{- else }}
+            {{- tuple . (coalesce .Values.dbName .Values.db_name) (coalesce .Values.dbUser .Values.global.dbUser "root" | include "resolve_secret_urlquery") (coalesce .Values.dbPassword .Values.global.dbPassword .Values.pxc_db.root_password | include "resolve_secret_urlquery" | required ".Values.pxc_db.root_password is required!") .Values.pxc_db.name | include "utils._db_url_pxc_db_v2" }}
+        {{- end }}
+    {{- else -}}
+mysql+pymysql://{{ include "db_credentials" . }}@
```
Comment (s10) [inline openstack/utils/templates/_hosts.tpl:171]: WIP: need to be adapted

File: openstack/utils/templates/snippets/_proxysql.cfg.tpl [hunk 2]
```diff
@@ -43,7 +44,7 @@ mysql_servers =
 {{- range $index, $dbKey:= $.dbKeys }}
```
Comment (s10) [inline openstack/utils/templates/snippets/_proxysql.cfg.tpl:44]: WIP: proxysql_monitor

**Review by galkindmitrii — APPROVED** [review]
lgtm, thanks for making it backwards compatible

File: openstack/utils/templates/_hosts.tpl [hunk 3]
```diff
@@ -24,38 +24,70 @@
...
+        {{- else if eq $dbType "pxc-db" -}}
+            {{- $dbConfig = $envAll.Values.pxc_db -}}
+        {{- end -}}
+        {{- if and $dbConfig.users $dbConfig.databases }}
+            {{- $db := first $dbConfig.databases }}
+            {{- $user := get $dbConfig.users $db | required (printf ".Values.db.%v.name & .password are required (key comes from first database in databases array)" $db) }}
+            {{- $user.name | required (printf ".Values.db.%v.name is required!" $db ) }}:{{ $user.password | required (printf ".Values.db.%v.password is required!" $db ) }}
         {{- else }}
-            {{- coalesce .Values.dbUser .Values.global.dbUser "root" }}:{{ coalesce .Values.dbPassword .Values.global.dbPassword .Values.mariadb.root_password | required ".Values.mariadb.root_password is required!" }}
+            {{- coalesce $envAll.Values.dbUser $envAll.Values.global.dbUser "root" }}:{{ coalesce $envAll.Values.dbPassword $envAll.Values.global.dbPassword $dbConfig.root_password | required ".Values.db.root_password is required!" }}
```
Comment (s10) [inline openstack/utils/templates/_hosts.tpl:42]: It looks like resolve_secret_urlquery wasn't added here, when we added resolve

File: openstack/utils/templates/_hosts.tpl [hunk 4]
```diff
@@ -24,38 +24,70 @@
 {{ include "_resolve_secret" (tuple . true) }}
 {{- end -}}
 
-{{define "db_url" }}
```
Comment (s10) [inline openstack/utils/templates/_hosts.tpl:27]: This function isn't being used anywhere, looks like a leftover after postgres->mariadb migration

**Review by businessbean — APPROVED** [review]
lgtm

File: openstack/utils/templates/_hosts.tpl (hunk 4 above)
Comment (fwiesel) [reply openstack/utils/templates/_hosts.tpl:27]: It is used in nova for the cells.

File: openstack/utils/templates/_hosts.tpl [hunk 5]
```diff
@@ -102,12 +161,70 @@ mysql+pymysql://{{ include "db_credentials" . }}@
...
+            {{- end }}
+        {{- end -}}
+        charset=utf8
+    {{- end }}
+{{- end }}
 
 # Please keep as it is, special case when it has to reference the db_region value.
-{{define "db_host_pxc"}}{{.Release.Name}}-percona-pxc.{{.Release.Namespace}}.svc.kubernetes.{{.Values.global.db_region}}.{{.Values.global.tld}}{{end}}
+{{- define "db_host_pxc" }}
+{{- .Release.Name -}}-percona-pxc.{{- .Release.Namespace -}}.svc.kubernetes.{{- .Values.global.db_region -}}.{{- .Values.global.tld -}}
```
Comment (fwiesel) [inline openstack/utils/templates/_hosts.tpl:207]: No need for the hyphens, when there is no whitespace in between.
```
{{- .Release.Name }}-percona-pxc.{{ .Release.Namespace }}.svc.kubernetes.{{ .Values.global.db_region }}.{{ .Values.global.tld }}
```
The `{{- end }}` below eats the whitespaces before.

In general, I'd follow the pattern only to "eat" the whitespaces on the left-hand side, unless there is a good reason not to (i.e. whitespaces for readability) `{{- ...}}` .

File: openstack/utils/templates/_hosts.tpl [hunk 6]
```diff
@@ -102,12 +161,70 @@ mysql+pymysql://{{ include "db_credentials" . }}@
...
+{{/*
+Custom DB URL for the global services using percona_cluster
+*/}}
+{{- define "utils._db_url_pxc_global" }}
+{{- $prefix := "mysql+pymysql" }}
+{{- $username := include "resolve_secret_urlquery" .Values.percona_cluster.db_user }}
+{{- $password := include "resolve_secret_urlquery" .Values.percona_cluster.dbPassword }}
+{{- $db_host := include "db_host_pxc" . }}
+{{- $db_name := .Values.percona_cluster.db_name }}
+{{- $prefix -}}://{{- $username -}}:{{- $password -}}@{{- $db_host -}}/{{- $db_name -}}?charset=utf8
```
Comment (fwiesel) [inline openstack/utils/templates/_hosts.tpl:226]: I'd rather see that as well in this form:
```
{{- $prefix }}://{- $username }}:{{ $password }}@{{ $db_host }}/{{ $db_name }}?charset=utf8
```

File: openstack/utils/templates/_hosts.tpl (hunk 4 above)
Comment (s10) [reply openstack/utils/templates/_hosts.tpl:27]: In the nova chart for the cells the `db_url_mysql` function is being used, not db_url directly.
db_url in the past was used conditionally if postgresql was enabled, and it's disabled for all OpenStack services now either implicitly, or explicitly with configuration:
```
postgresql:
  enabled: false
```

Comment (fwiesel) [reply openstack/utils/templates/_hosts.tpl:27]: Ah, no. I am mistaken.

**Pull request merged**